Option to save default layout workbench

Harrietbrown opened the pull request:
**Description of work.**
I have implemented an option to save the current layout of the workbench widgets.
On startup workbench will load widgets in the state they were when last closed.
Layout can be reset to any user to using option under view.

**To test:**
1. Open Workbench.
1. Change widget dock layout.
1. ``View --> User Layouts --> Settings``
1. Save your layout.
1. Test that the layout you just saved is accessable in the settings lesection box.
1. Check that your saved layout is listed under ``View --> User Layouts``
1. ``View --> Restore default layout``
1. Test that the option to restore your layout works from the settings menu and from ``View --> User Layouts --> [your layout name]``.
1. Change widget dock layout.
1. Restart Workbench.
1. Check that the the Workbench layout restores the last layout state.
1. Check that your layout is still available.
1. Delete your layout from the settings menu
1. Check that the layout is removed from the settings selection box and ``View --> User Layouts``.

Fixes #25515.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Harrietbrown committed:
First Commit, add option to set layout to alternate default

Harrietbrown committed:
Implement option to store current layout to user configs

Option to store and load layout included under view on the toolbar

Harrietbrown committed:
workbench startup loads user default if one provided.

Harrietbrown committed:
Add new feature to version 4.1.0 resesase notes.

Harrietbrown committed:
Correct formatting for propper coding practice.

Harrietbrown committed:
Update qt/applications/workbench/workbench/app/mainwindow.py

hsautessella commented on qt/applications/workbench/workbench/app/mainwindow.py:
```diff
@@ -416,21 +425,27 @@ def add_dockwidget(self, plugin):
+        CONF.set("Layout/Layout", self.saveState())
+
+    def load_layout_settings(self):
+        """Load a layout for the child widgets from the user config, if no user default exists load the default"""
```
This doc-string is a bit long, can we split it into two lines?

Harrietbrown replied:
```suggestion
        """Load a layout for the child widgets from the user config, 
        if no user default exists load the default"""
```

hsautessella commented on qt/applications/workbench/workbench/app/mainwindow.py:
```diff
@@ -416,21 +425,27 @@ def add_dockwidget(self, plugin):
+    def load_layout_settings(self):
+        """Load a layout for the child widgets from the user config, if no user default exists load the default"""
+        try:
+            self.restoreState(CONF.get("Layout/Layout"))
```
I know it was my suggestion to create a `[Layout]` section, but looking at the .ini file there's an existing `[MainWindow]` section. Perhaps it makes sense to save this to `MainWindow/UserLayout` or something like that?

Harrietbrown replied:
```suggestion
            self.restoreState(CONF.get("MainWindow/UserLayout"))
```

Harrietbrown commented on qt/applications/workbench/workbench/app/mainwindow.py:
```diff
@@ -416,21 +425,27 @@ def add_dockwidget(self, plugin):
-        self.setup_default_layouts()
+    def save_layout_settings(self):
+        """Save the current layout to configuration"""
+        CONF.set("Layout/Layout", self.saveState())
```
```suggestion
        CONF.set("MainWindow/UserLayout", self.saveState())
```

Harrietbrown committed:
Apply suggestions from code review

Harrietbrown commented on qt/applications/workbench/workbench/app/mainwindow.py:
```diff
@@ -416,21 +425,28 @@ def add_dockwidget(self, plugin):
+        CONF.set("MainWindow/UserLayout", self.saveState())
+
+    def load_layout_settings(self):
+        """Load a layout for the child widgets from the user config, 
```
```suggestion
        """Load a layout for the child widgets from the user config,
```

Harrietbrown committed:
Update qt/applications/workbench/workbench/app/mainwindow.py

Harrietbrown commented on qt/applications/workbench/workbench/app/mainwindow.py:
```diff
@@ -300,10 +300,19 @@ def create_actions(self):
+            on_triggered=self.setup_default_layouts,
             shortcut="Shift+F10", shortcut_context=Qt.ApplicationShortcut)
+        action_save_user_layout = create_action(
+            self, "Set Custom Layout",
```
```suggestion
            self, "Save Custom Layout",
```

Harrietbrown committed:
Update qt/applications/workbench/workbench/app/mainwindow.py

martyngigg commented:
Quick question: There was already a `self.saveState` line going into the config [here](https://github.com/mantidproject/mantid/blob/95e717264b3080dd0e88713e0f50fb52f03a0675/qt/applications/workbench/workbench/app/mainwindow.py#L649) do we know why that wasn't restored properly by [readSettings](https://github.com/mantidproject/mantid/blob/95e717264b3080dd0e88713e0f50fb52f03a0675/qt/applications/workbench/workbench/app/mainwindow.py#L636)?

The issue here was actually slightly poorly described by myself. I thought the workbench already saved/reloaded any layout changes anyway without the user having to do anything (apparently not but if we could investigate the above question). The request was to be able to define named layouts that people could save and reset to them. So extending this slightly to save ask for a name and put this as an extra option under the "reset to layout" option.

Harrietbrown commented:
The state is loaded before the widgets are set in the layout, so any previous layout is overwritten. It can be fixed by changing the order of functions, but I think it is better to load the user layout instead of the state when it was last closed.

Harrietbrown committed:
Make saving multiple user layouts available possible through settings

Workbench no longer loads the user set default on startup instead restoring the
factory default. load and save user layout have been removed from view menu and
moved to the settings. loading settings user settings are available from a
submenu in view.

Harrietbrown committed:
fix cases of dict.keys() to list(dict.keys())

martyngigg commented on qt/applications/workbench/workbench/widgets/settings/general/section_general.ui:
```diff
@@ -305,6 +305,42 @@
+             </widget>
+            </item>
+            <item row="3" column="0" colspan="3">
+             <widget class="QLineEdit" name="new_layout_name"/>
```
Maybe add a label on the left of the text box saying "Layout name:"?

martyngigg reviewed (changes requested):
This is looking nice. I think we could add some release notes.

Some errors I ran into:

If I click either of the load/delete layout buttons when nothing is selected in the layouts box then I get:
```
  File "/media/data1/source/github/mantidproject/mantid/qt/applications/workbench/workbench/widgets/settings/general/presenter.py", line 166, in delete_layout
    layout = self.view.layout_display.currentItem().text()
AttributeError: 'NoneType' object has no attribute 'text'```

martyngigg commented:
I also note that it doesn't yet automatically reload the layout as it was in the last session. Can we do that? Ideally users shouldn't need to force a save of a user layout to get the layout they changed last time. It's only intended for saving multiple named layouts should users need it.

martyngigg commented:
I'm now out for a week so I haven't assigned myself for further testing here.

Harrietbrown committed:
Fix bug where load/delete would cause error if no item selected

Harrietbrown committed:
Change layout setup to reload latour last session

Harrietbrown committed:
Add tests for layout loading functions

Harrietbrown committed:
Removed black lines at end of test file

hsautessella commented:
Do you think it would be a good idea to include a link to the layout settings inside the `User Layouts` subwindow? It feels odd having an empty subwindow if you haven't got any layouts saved and it's not immediately obvious where to add items to it. Possibly something like this? Perhaps `Save/Load Layouts` would be a better label than `Show Layout Settings`.

![image](https://user-images.githubusercontent.com/46960113/60173634-17066980-9807-11e9-99e7-eb4740a99721.png)

Maybe we could add an optional `focus_widget` argument to `MainWindow.open_settings_window` to take the user straight to the relevant settings.

Harrietbrown committed:
Add link to settings in user settings in view

Harrietbrown commented:
I've added the settings, but given that this is kind of a specific case I have made a seperate action to load settings instead of adding keywords to the existing one. If there are going to be several cases where we need to link to specific sections within settings then it will be worth making it all a single function, but that is going to need to integrate switching between sub-settings which we don't have yet so it's not really worth setting it up with that right now.

hsautessella commented on qt/applications/workbench/workbench/app/mainwindow.py:
```diff
@@ -554,6 +581,11 @@ def open_settings_window(self):
+    def open_settings_layout_window(self):
+        settings = SettingsPresenter(self)
+        settings.general_settings.focus_layout_box()
+        settings.show()
```
For me the `new_layout_name` box isn't actually made visible by this function. This is what I see after clicking `View -> User Layouts -> Settings`:
![image](https://user-images.githubusercontent.com/46960113/60272897-e7319180-98ec-11e9-8537-3f9f20863664.png)

I think this is because Qt doesn't actually know how far down to scroll until the widgets are shown, so swapping lines 586 and 587 should help.

hsautessella replied:
On my screen half the `Layout Name` box is still hidden when showing the layout settings.
![image](https://user-images.githubusercontent.com/46960113/60329489-5c9e7000-9988-11e9-9d3c-0285ed966566.png)

 There's an optional `yMargin` argument in `ensureWidgetVisible`. From my testing setting this to 75 works on my machine (the default is 50).

Harrietbrown committed:
Apply changes from review

Harrietbrown committed:
Layouts box in setting is visible when accessed from view

martyngigg commented:
@Harrietbrown This looks good. I'll merge it now. I think it would be good to add some documentation to the [workbench](https://github.com/mantidproject/mantid/tree/master/docs/source/workbench) section. It looks like we are missing a page on the settings dialog. Would you mind creating a new issue to do that?

Harrietbrown commented:
Sure thing, are we going to want to have that documentation done for this release?

martyngigg commented:
If possible that would be helpful.
